### DISCUSSION TITLE
New Navidrome Web Player Proposal

### DISCUSSION THREAD

**Discussion by deluan** [navidrome/navidrome#2285] [Ideas]
# Navidrome Web Player Requirements

## Introduction
In light of the [deprecation](https://github.com/navidrome/navidrome/issues/1651) of the [React Music Player](https://github.com/lijinke666/react-music-player), it has become necessary to find a new solution for Navidrome's music player. This requirements document aims to outline the essential components needed for developing a new web player that will address long-standing issues and provide better overall functionality. The new player should be built on a React/Material-UI framework, compatible with most browsers and media formats, and easily extensible.

## 1. Background
Navidrome is currently using its own fork of the React Music Player with a couple of bug fixes applied to it. The decision has been made to develop a new player to better suit the needs of the project. Some previous discussions about this can be found on our Discord (ex: https://discord.com/channels/671335427726114836/704303730660737113/861006516315881522)

## 2. Requirements
The new music player should meet the following criteria:

### 2.1. Written in React/Material-UI
The player must be developed using the React/Material-UI framework to seamlessly blend into Navidrome's existing UI and support theming.

### 2.2. Browser and Format Compatibility
The player must be compatible with most, if not all, browsers and support the main media formats. It should be based on the HTML5 Media Player.

### 2.3. Gapless Playback
The player must support gapless playback, as described in issue #745.

### 2.4. Normalization/ReplayGain Support
The player must include normalization and ReplayGain support, as discussed in issue #233.

### 2.5. Remote API Control
The player must have the capability to control the remote API for Jukebox, as described in issue #364.

### 2.6. Internet Radio Integration
The player should integrate well with the Internet Radio feature, displaying the current song being played if the information is available.

### 2.7. Extensibility
The player's UI should be easy to extend, allowing for the addition of more options in the future.

### 2.8. Internationalization
The player must support internationalization, allowing for translations and localization to cater to users from different regions and languages.

### 2.9. Accessibility
The player should be designed with accessibility in mind, ensuring that it is usable by individuals with disabilities.

### 2.8. Separation as an Independent Project
The player may be developed as a separate project, if necessary.

## 3. Implementation Details
The initial idea is to create a React player using the [useAudioPlayer](https://github.com/E-Kuerschner/useAudioPlayer) library, which allows the specification of an audio provider. This would enable an HTML5 audio provider for playback in the browser and a remote audio provider for controlling the Jukebox. However, it is unclear if the useAudioPlayer library supports gapless playback. If not, a workaround could involve creating a new provider that wraps the [Stitches library](https://github.com/sudara/stitches).

## 4. Conclusion
This requirements document outlines the necessary components for the development of a new Navidrome music player. The player should be built using the React/Material-UI framework and be compatible with most browsers and media formats. It must support gapless playback, normalization, ReplayGain, remote API control, Internet Radio integration, extensibility, accessibility, and internationalization. Implementation using the useAudioPlayer library is proposed, with a potential workaround for gapless playback using the Stitches library if needed.

**Comment by andrewdunndev** [comment]
This _likely_ would be out of scope, but proposing for sake of discussion (potentially this suggestion needs to be moved into #364 or be it's own discussion):

> 2.5. Remote API Control

Within the context of #364 where the desire is to set a "playback sink", I wonder if it would be instead more simple to create a sort of "navidrome satellite" which would be a copy of this web player paired up with something like mpv (with its [`audio-exclusive`](https://mpv.io/manual/master/#options-audio-exclusive) mode). Similar to what clients like [feishin](https://github.com/jeffvli/feishin) are, but simpler to deploy as a headless (e.g. on a small device connected to a USB DAC) playback system that you control over http.

**Reply by dweymouth** [reply]
If we go this route I'd be happy to split out the mpv player component for Supersonic into its own repo so it can be used as a base for this. (It dynamically links against libmpv.) But maybe a different/better(?) idea is to implement DLNA and/or Chromecast as a sink so the Navidrome web UI can cast to existing smart speakers, Chromecast devices, etc, and leverage existing projects like [this](https://github.com/hzeller/gmrender-resurrect) for the Raspberry PI case.

**Reply by andrewdunndev** [reply]
It's highly preferential, but I'd vote to prioritize simpler playback (with exclusive for sample rate switching) above integration with smart devices. This would allow _all_ peripheral speakers to be utilized rather than a specific ecosystem.

**Reply by dweymouth** [reply]
Were you envisioning the "Navidrome satellite" being deployed on different hardware than the Navidrome server? It's definitely a cool idea but I think it would have to advertise itself on the LAN so the server and/or device running the web client can be aware of its existence, at which point you're basically re-inventing uPnP/DLNA, which is why I'd mentioned the idea of using an existing standard.

Jukebox mode is certainly the easiest path forward, with the remote playback node being part of the server itself (and you'd need to deploy the server on the same device connected to your audio system)

Hmm though I guess you could have the remote playback node be configured with the server's connection details, and have an always-open connection to the server, and the server could then relay control messages from the web client.

**Reply by andrewdunndev** [reply]
> Were you envisioning the "Navidrome satellite" being deployed on different hardware than the Navidrome server?

Yes, so it's a bit _more/different_ than #364 in its current form.

> It's definitely a cool idea but I think it would have to advertise itself on the LAN so the server and/or device running the web client can be aware of its existence, at which point you're basically re-inventing uPnP/DLNA, which is why I'd mentioned the idea of using an existing standard.

Instead of it being a "sink" like uPnP/DLNA, I'd thought of it more as a slim client that just pulls the common portions from the navidrome codebase so you have a similar look and feel. Instead of playing to the browser though it would play to some locally configured playback device (I'd assume default to mpv).

This would be to hit the "headless playback" use cases that the community of people using [mpd](https://www.musicpd.org/) where they typically want to play via an external DAC and play "bit perfect" (e.g. switching sample rate, PCM, DSD). 

> Hmm though I guess you could have the remote playback node be configured with the server's connection details, and have an always-open connection to the server, and the server could then relay control messages from the web client.

Maybe it's simpler to do something like this where instead of trying to be compliant with uPnP/DLNA you just make a small mpv+shim that can be deployed someplace and registered as a player with the navidrome server.

The main goal in my suggestion is:
 - control of playback from web, on a device like a phone/tablet
 - playback via another device, via USB dac
 - audio exclusive playback of PCM (sample rate switching), bonus for playback of DSD (currently just converts to PCM within navidrome)

**Reply by SteveDinn** [reply]
You guys are the best. I found a couple of Raspberry Pi Zeroes on my desk and I was wondering if there was something out there like this. Google brought me to a 2-year-old Reddit thread, which brought me to an already-closed issue about Jukebox mode, which brought me here :)

I would love to have something that would run on my RPi Zeroes that would:
- act as a Subsonic client app that is accessible via HTTP
- preferably use Navidrome's web UI
- play locally on the (client) device, like Jukebox mode.
- get its metadata and streams from a configurable 'principal' Subsonic server
    - the server doesn't have to be discoverable, I would be fine with entering the URL and credentials manually

It could even pull the web app from some Navidrome-adjacent location so that I never have to touch the RPis again.

@andrewgdunn, I think we're almost describing the same thing.

Basically, I want somebody (a guest in my house, perhaps) to be able to pick up a tablet that I have on my coffee table, open the PWA for this web client, and be able to play music on my stereo by way of the RPiZero where this client is running.

I think this is a bit beyond the scope of the new web-player component discussion though :)

**Comment by hboetes** [comment]
Would this player suit the bill (mostly)? https://cookpete.com/react-player/

**Reply by bilogic** [reply]
I tested https://coderadio-admin-v2.freecodecamp.org/listen/coderadio/radio.mp3 and did not show the title

**Reply by hboetes** [reply]
That's because it's playing straight in your browser. ;-) Check the test things on that page, or even better, read the GitHub link at the bottom to learn how to use it.

**Reply by bilogic** [reply]
You seem to suggest that after reading the link and not playing music straight in the browser, it will be able to show its title and address item 2.6. 

Given this is meant to replace Navidrome's player, where should I load this player if not in a browser?

**Comment by sammbeller** [comment]
After discussion with Deluan it looks like we're going to be building our own solution based in part on `useAudioPlayer` which doesn't appear to be supported. Before I dig too deep into a proposed solution I was hoping we could solidify our understanding of the API I'll be implementing against

# `useAudioPlayer` current API
The current `useAudioPlayer` API looks like this (at least when using the recommended `useGlobalAudioPlayer`
```ts
export interface AudioPlayerState {
    src: string | null
    looping: boolean
    isReady: boolean
    isLoading: boolean
    paused: boolean
    stopped: boolean
    playing: boolean
    duration: number
    muted: boolean
    rate: number
    volume: number
    error: string | null
}

export interface UserListeners {
    onstop?: () => void | undefined
    onpause?: () => void | undefined
    onload?: () => void | undefined
    onend?: () => void | undefined
    onplay?: () => void | undefined
}

export interface AudioLoadOptions extends UserListeners {
    loop?: boolean
    autoplay?: boolean
    initialVolume?: number
    initialMute?: boolean
    initialRate?: number
    format?: string
    html5?: boolean
}

export type LoadArguments = [src: string, options?: AudioLoadOptions]

export interface AudioPlayer extends AudioPlayerState {
    play: () => void
    pause: () => void
    togglePlayPause: () => void
    stop: () => void
    setVolume: (volume: number) => void
    fade: (from: number, to: number, duration: number) => void
    setRate: (speed: number) => void
    seek: (seconds: number) => void
    mute: (muteOnOff: boolean) => void
    loop: (loopOnOff: boolean) => void
    getPosition: () => number
    load: (...args: LoadArguments) => void
}
```

# Proposed API
I propose we expose an API that looks like this:
```ts
enum AudioPlayerState {
    LOADING: "LOADING",
    PAUSED: "PAUSED",
    PLAYING: "PLAYING",
    STOPPED: "STOPPED",
}

export interface LoadOptions {
    // TODO should we include callbacks?
    ...AudioPlayerState
}

export interface AudioPlayer {
    error?: string;
    getDuration: () => number;
    getIsMuted: () => boolean;
    getIsSetToLoop: () => boolean;
    getPlaybackRate: () => number;
    getPosition: () => number;
    getState: () => AudioPlayerState;
    getVolume: () => number;
    load: (src: string, options: LoadOptions) => void;
    play: () => void;
    pause: () => void;
    setFade: (from: number, to: number, duration: number) => void;
    setIsMuted: (isMuted: boolean) => void;
    setIsSetToLoop: (isSetToLoop: boolean) => void;
    setPosition: (position: number) => void;
    setPlaybackRate: (playbackRate: number) => void;
    setVolume: (volume: number) => void;
    stop: () => void;
    src?: string;

}
```
## Rationale for changes
The `AudioPlayerState` is tightly coupled to the `howler` API and, additionally, duplicates a lot of state. Rather than duplicate this state, I expose getter and setter functions which will call into functions exposed by the `AudioOutProvider` the API of which will need to support both `howler` and the subsonic API for navidrome. Given that, a lot of these probably need to be async.

## Open questions
One of the big open questions is whether or not we do the work to expose the callbacks currently supported by `useAudioPlayer` via `LoadArguments`. Please let me know what you all think

**Reply by sammbeller** [reply]
An additional concern to consider wrt the asynchronous nature of jukebox mode is that if we aim to include the callbacks that we'll need either some sort of polling system or to implement webhooks.

**Reply by deluan** [reply]
I think your proposed API makes more sense than the original one, and yes, we will need the api to be async and have callbacks. I'll add a WebSocket to get events from the server.

Just a small style nitpick, I prefer not to use getter/setter style. e.g: `isMuted`/`mute` instead of `getIsMuted`/`setIsMuted`

**Reply by deluan** [reply]
Another consideration is that we have the need to use this API to control both devices that maintain their own queue (ex: Jukebox API) and that only play one file at a time (ex: howler). Not sure how to best accommodate this.

**Reply by sammbeller** [reply]
> Just a small style nitpick, I prefer not to use getter/setter style. e.g: `isMuted`/`mute` instead of `getIsMuted`/`setIsMuted`

These are functions and asynchronous functions at that. I think we need some way to indicate by  name that these functions will be doing some work and `isMuted` to me is just the name of a variable. Obviously at the end of the day this isn't a big deal but I kind of feel strongly about this.

> Another consideration is that we have the need to use this API to control both devices that maintain their own queue (ex: Jukebox API) and that only play one file at a time (ex: howler). Not sure how to best accommodate this.

There are a bunch of solutions here not least of which is a simple flag returned indicating whether or not the audio player supports queuing. I'll probably leave this for last though for this reason.

**Comment by sammbeller** [comment]
The work for this project can be tracked [here](https://github.com/sammbeller/navidrome-audio-player)

**Comment by wu-lee** [comment]
Hello! I am interested in being able to control the jukebox via the UI, therefore I've been watching this discussion and the above repo for quite a while.  Can you let me know what the status of this is, and if there is any way to advance it, perhaps with contributions?  Thanks!

**Reply by sammbeller** [reply]
Bumping this @deluan

**Comment by zhaodawei** [comment]
I’m using Navidrome on a Linux NAS with a USB DAC. Jukebox playback through MPV works well, but I currently need Feishin to control it.
I would love to control server-side playback directly from Navidrome’s Web UI, including play/pause, seeking, and queue management, so that any phone or computer could act as a remote without installing another client.
A related improvement would be an admin page to discover and select server-side audio devices. I switch between USB DACs, and currently have to edit Jukebox.Devices / Jukebox.Default and restart Navidrome.
Is Web UI support for Jukebox still planned, and should audio-device management be tracked as a separate feature request?
